### ISSUE TITLE
Plugins Sourced in MinVR?

### ISSUE THREAD

**Issue by dtorban** [ivlab/MinVR2#13]
I was thinking of sourcing some of the plugins inside of MinVR itself in the contrib/plugins directory.  Here are the plugins I was thinking of sourcing:
- vrpn
- glut
- glfw

Any others?  None?  Where should they be sourced?  I was thinking of adding them to the contrib directory because they are not necessary for the core of MinVR.  The plugins would download and install their own dependencies.

**Comment by dankeefe** [comment]
This is a very good question.  I lean toward what you suggest, but here are the pros/cons I can think of...  I guess there are 3 ways we can go in terms of a policy on where plugins should live:
1. All plugins are sourced outside of MinVR.
2. All are sourced inside MinVR.
3. Most are sourced outside of MinVR, but a couple of the most important (vrpn, glut and/or glfw -- maybe one for graphics and one for input) are sourced inside MinVR.

I think I would lean toward option 3 as you suggest because these plugins could serve as a useful example of how to write a plugin.  We could say for example that VRPN and glut are "officially supported" plugins that we all design together with the idea that they serve as the model for how to write a good plugin.  We would still not want the main MinVR library do depend on the plugins since they in turn depend on other libraries, so I guess what we would do in this situation is to make it an option to build or not build the example plugins when you configure the MinVR build system.  Or, maybe better, have a totally separate build system for these plugin directories.  So, that even the build system acts as an example that others can copy in order to make their own plugins.

Considering the other options:

Option 2 (everything inside MinVR) does not seem good because the repo will get bloated and confusing.

Option 1 (everything outside MinVR) seems like it could work if we have some way to make it very, very easy for people to find example plugins.  This could be handled most easily via a MinVR webpage that makes it easy to find and download plugins.  Or, maybe it could be handled via functionality in git or cmake that would help with nesting one project's source inside another -- I don't know too much about this, but seems like Dan O. has already had some success with this.  So, I could also buy into this option if people feel strongly.

**Comment by BenKnorlein** [comment]
I would even go for something similar to Option 2 so that plugins are collected and built with MinVR if needed. This does not necessarily mean all plugins, but at least the ones which we consider essential and useful. Here are the reasons why i think this would be the best way.
- It makes it easier for people to see what is out there and they can check what they can use. Finding plugins at different places might be difficult.
- It provides better compatibility in case there are things which change the interface or somewhere else, In this case it is easier to ensure that plugins are compatible with the core
- It is also easier to develop basic tests, which at least ensure a certain stability for MinVR and the components.

I think the best would be to have them as submodules, so that they all have their own repo https://git-scm.com/book/en/v2/Git-Tools-Submodules

With CMake we can then control which plugins are built, Maybe it is even possible to only download the source of plugins when building/configuring (I am not 100% sure if this is possible when we use submodules).

So to summarize for me it could look something like this:
- Each plugin has its own repo and is included as a submodule in the MinVR2 repo
- The build process for the plugin is supported and controlled by the MinVR2 build process, so that it does not have to be compiled by its own.
- The source is only downloaded when the plugin is enabled during compile time.

**Comment by dankeefe** [comment]
- I really like the git submodules approach for at least a set of "commonly used" plugins.  Submodules seem tailor made for this.
- But, not all plugins will be able to be made submodules.  I'm imagining that there will be many, many plugins and some will never get committed to the main MinVR repo because they are experimental or depend on some code that can't be made public or for some other reason.  
- For the non-submodule situation, I'd like to be able to work on one of these plugins inside the MinVR/contrib (or similar) directory and be able to build it and experiment without making a change to a MinVR/CMakeLists.txt.  It would be nice if that could just be stable.
- Can we support this in some way?  Maybe cmake automatically picks up and creates a new rule to build a plugin for all of the subdirectories that it can find inside MinVR/contrib?

**Comment by tsgouros** [comment]
Won't cmake just delegate the build of whatever is in MinVR/contrib to the CMakeLists.txt file it finds in there? Is there something more that you're after?

**Comment by BenKnorlein** [comment]
Yes. if you include a folder it will just call the CMakeLists.txt in that folder.

**Comment by dankeefe** [comment]
I'm asking two things I think:
1. Can we avoid modifying the root CMakeLists.txt to add the line
   "add_subdirectory(contrib/myplugin") for every plugin?  Maybe CMake can
   search for existing directories and automatically add them or something
   like that.  What I'm trying to avoid here is having every user need to
   modify the root CMakeLists.txt and then accidentally committing changes to
   add or disable whatever plugins they are using on their local system to the
   main repo.  Maybe there is another strategy like a local build config file
   or CMake command line options that would be better than automatically
   adding all contrib subdirectories?  Just want to avoid accidental changes
   to the main CMakeLists.txt based on plugins used at one site but not
   others.
2. Would building a MinVR plugin require placing it together with the main
   MinVR project so that it will correctly inherent CMake settings for naming
   conventions or #defines or other things from the root CMakeLists.txt file?
   Or since the plugins are technically stored in their own repos would we
   want to be able to build them separately without MinVR?  Actually, I guess
   even in a separate build they would need to link with the MinVR plugin code
   so maybe it's not a limitation to require that you just build within the
   MinVR source tree.  That sounds ok to me the more I think about it.

On Wednesday, December 16, 2015, BenKnorlein notifications@github.com
wrote:

> Yes. if you include a folder it will just call the CMakeLists.txt in that
> folder.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/ivlab/MinVR2/issues/13#issuecomment-165254170.

## 

Daniel Keefe
Associate Professor
Department of Computer Science and Engineering
University of Minnesota
http://www.cs.umn.edu/~keefe/

**Comment by tsgouros** [comment]
I think you can just say add_subdirectory(contrib) and leave the
CMakeLists.txt file in that subdirectory to figure it out. That only
delegates the same problem to the lower level, but it will keep the main
cmake file from being polluted.

Or there's this:
http://stackoverflow.com/questions/7787823/cmake-how-to-get-the-name-of-all-subdirectories-of-a-directory
which seems to be a pretty smooth solution to the 1.
1. It is definitely a limitation we want to avoid to force plugins to build
   within the MinVR source tree. They should be able to compile with an
   include file and the location of the library. And of course at runtime they
   should be installed somewhere that MinVR knows to look for them.

**Comment by BenKnorlein** [comment]
To 1) I think the Macro is a nice way to allow easy integration of plugins, so I think that this should be the way for the contrib folder

To 2) I think the easy and preferred way for a user should be to just add the plugins he wants to use to the contrib folder and the MinVR2 build process will take care for it.

If a user does not want to do this he has to link to the headers and a library as Tom already wrote. To facilitate this we could provide a "FindMinVR.cmake" script which developer can use within their CMakeFiles.txt with find_package(MinVR). We could even allow defining modules, e.g. find_package(MinVR core vrpn). I think we need the Find script anyways for applications.

**Comment by dankeefe** [comment]
Which macro?

**Comment by dtorban** [comment]
I think perhaps using CMake external projects (https://cmake.org/cmake/help/v3.0/module/ExternalProject.html) is a better approach than using submodules.  I have had quite a bit of success with the external project functionality.  Here are some reasons:
- Submodules tie your project tightly to a version of the external project.  In other words, anytime someone makes a change to the plugin, we will also have to update MinVR.  This could be a maintenance headache.
- CMake External projects allow you to monitor a branch more loosely, so we don't have to update MinVR on every plugin change.  We can also tightly couple if we want to, by monitoring a version.
- CMake external projects integrate very well into the cmake build process:
  - Downloads the project if included
  - Automatically builds the project using the current configuration
  - Allows you to specify dependancies, etc...
  - You can specify where it downloads the project and where the build directory is.
  - It can automatically update and patch changes from the external project if you want it to.
- Submodules are weird because you have to manually tell git to download them, and I don't think they integrate very well into cmake downloading them.  I recall looking at this in the past.

Overall, I think we have much more power and flexibility with using external projects vs submodules.  Let me know your thoughts.  For things we want to tightly couple to, submodules may be a good approach, but we may have problems integrating it well into the cmake build system.

Thanks,
Dan

**Comment by dtorban** [comment]
Here is an example in MinVR of using a cmake external project.  It is pretty easy:
https://github.com/ivlab/MinVR2/blob/master/contrib/CMakeLists.txt

**Comment by BenKnorlein** [comment]
cmake external project sounds nice. I am not really familiar with it but will look into it. I also became a little bit unsure if submodules are the right thing the more i thought about it.

And the Macro is in the link Tom added.
http://stackoverflow.com/questions/7787823/cmake-how-to-get-the-name-of-all-subdirectories-of-a-directory

**Comment by dankeefe** [comment]
I see.  So, it sounds like this is in part a question of whether we organize around cmake or git.  Good points, Dan O.  Thanks.

**Comment by tsgouros** [comment]
The build for the Threaded and GLFW plugins is broken. Both have C++ errors (uninstantiated objects, empty forward declarations, etc).

**Comment by dtorban** [comment]
I opened a new issue for this one.

**Comment by dtorban** [comment]
Closed, since I think we have a handle on this.
